lib/spack/spack/cmd/external.py

```diff
@@ -0,0 +1,271 @@
+
+    For convenience, this is constructed as a dictionary where the keys are
+    the executable paths and the values are the names of the executables
+    (i.e. the basename of the executable path).
+
+    There may be multiple paths with the same basename. In this case it is
+    assumed there are two different instances of the executable.
+    """
+    path_hints = spack.util.environment.get_path('PATH')
+    search_paths = llnl.util.filesystem.search_paths_for_executables(
+        *path_hints)
+
+    path_to_exe = {}
+    # Reverse order of search directories so that an exe in the first PATH
+    # entry overrides later entries
+    for search_path in reversed(search_paths):
+        for exe in os.listdir(search_path):
+            exe_path = os.path.join(search_path, exe)
+            if is_executable(exe_path):
+                path_to_exe[exe_path] = exe
+    return path_to_exe
+
+
+ExternalPackageEntry = namedtuple(
+    'ExternalPackageEntry',
+    ['spec', 'base_dir'])
+
+
+def _generate_pkg_config(external_pkg_entries):
+    """Generate config according to the packages.yaml schema for a single
+    package.
+
+    This does not generate the entire packages.yaml. For example, given some
+    external entries for the CMake package, this could return::
+
+       { 'paths': {
+             'cmake@3.17.1': '/opt/cmake-3.17.1/',
+             'cmake@3.16.5': '/opt/cmake-3.16.5/'
+         }
+       }
+    """
+    paths_dict = syaml.syaml_dict()
+    for e in external_pkg_entries:
+        if not _spec_is_valid(e.spec):
+            continue
+        paths_dict[str(e.spec)] = e.base_dir
+    pkg_dict = syaml.syaml_dict()
+    pkg_dict['paths'] = paths_dict
+
+    return pkg_dict
+
+
+def _spec_is_valid(spec):
+    try:
+        str(spec)
+    except spack.error.SpackError:
+        # It is assumed here that we can at least extract the package name from
+        # the spec so we can look up the implementation of
+        # determine_spec_details
+        tty.warn('Constructed spec for {0} does not have a string'
+                 ' representation'.format(spec.name))
+        return False
+
+    try:
+        spack.spec.Spec(str(spec))
+    except spack.error.SpackError:
+        tty.warn('Constructed spec has a string representation but the string'
+                 ' representation does not evaluate to a valid spec: {0}'
+                 .format(str(spec)))
+        return False
+
+    return True
+
+
+def external_find(args):
+    if args.packages:
+        packages_to_check = list(spack.repo.get(pkg) for pkg in args.packages)
+    else:
+        packages_to_check = spack.repo.path.all_packages()
+
+    pkg_to_entries = _get_external_packages(packages_to_check)
+    _update_pkg_config(pkg_to_entries, args.not_buildable)
+
+
+def _group_by_prefix(paths):
+    groups = defaultdict(set)
+    for p in paths:
+        groups[os.path.dirname(p)].add(p)
+    return groups.items()
```

Comment:
I think we can return:
```python
result = sorted([(prefix, sorted(s)) for prefix, s in groups.items()])
```
to solve the TODO below

Reply:
That would do it, although the TODO I have in the code mentions that it might be better to do it based on order of prefixes in `PATH` (so the results are repeatable and follow an intuitive rule).